### ISSUE TITLE
no error when using `+` operator on incompatible types

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#48962]
# Bug Report

today i was shocked to learn that something like `1 + "2"`, one of the most common examples of why javascript is unsafe, doesn't cause an error in typescript.

i get that it technically works at runtime, but so does `1 + undefined` which correctly causes an error

related: #30239

### 🔎 Search Terms
concatenate incompatible types operator

(i'd be very surprised if this hasn't been raised before, but i couldn't find an existing issue no matter what search terms i used)

### 🕗 Version & Regression Information
4.7.0-dev.20220504

### ⏯ Playground Link
https://www.typescriptlang.org/play?ts=4.7.0-dev.20220504#code/IwAg1CBEBMlA
### 💻 Code

```ts
1 + "2"
```

### 🙁 Actual behavior
no error

### 🙂 Expected behavior
error when using `+` operator on incompatible types

**Comment by fatcerberus** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/16895.

**Comment by MartinJohns** [comment]
> something like 1 + "2", one of the most common examples of why javascript is unsafe, doesn't cause an error in typescript.

Why would that be one of the most common examples of why JavaScript is unsafe? It's an obvious operation. It works in C# like this as well.

**Comment by fatcerberus** [comment]
Yeah, this is allowed because it’s common to write things like `numThings + " things"` in JS (less so with template literals, but it still happens enough that breaking it is undesirable).  TS doesn’t consider it unsafe adding numbers and strings because the behavior is completely predictable (it’s always a string concatenation, regardless of which side the number is on).

**Comment by DetachHead** [comment]
imo you would typically use string interpolation for cases like that. the languages i'm used to seem to make the same assumption.

### [python (mypy)](https://mypy-play.net/?mypy=latest&python=3.10&gist=33c13649e58cbf363f924aadcde98ac6)
```py
1 + "2" # error
f"{1}2" # no error
```
### [dart](https://dartpad.dev/?id=d614c846c43a9334ef8a5872cca75116)
```dart
void main() {
  1 + "2"; // error
  "${1}2"; // no error
}
```

### [kotlin](https://pl.kotl.in/gvuAn9VJh)
```kotlin
fun main() {
    1 + "2" // error
    "${1}2" // no error
}
```

**Comment by KotlinIsland** [comment]
### [kotlin](https://pl.kotl.in/5CMaUCIFZ)
```kts
1 + "2" // error
"1" + 2 // no error
```

Kotlin moment

**Comment by KotlinIsland** [comment]
I agree that the current behavior isn't optimal, I'd much prefer an error on `number + string` (and `string + number`), in interpolation it's obvious to the user that the object will be transformed into a string, but for normal `+` it's a bit sus.

**Comment by Josh-Cena** [comment]
Python is different: Python is a strongly typed language. Everything needs explicit casts. JavaScript, C, Java, etc. are all weakly typed.

**Comment by KotlinIsland** [comment]
Python and Javascript aren't that different, the main differences are mainly that Javascript tries to cover up mistakes a little a little more aggressively than Python:

```py
[1, 2] + 3 # error
class A:
    pass
A().a  # error
type(A())  # <class A>
```
```js
[1, 2] + 3 // no error, "1,23"
class A {}
new A().a  // no error, undefined
new A().constructor // class A{} 
```

What makes you think that Python is strongly typed?

**Comment by KotlinIsland** [comment]
> Because... that's literally the [definition of strongly typed](https://en.wikipedia.org/wiki/Strong_and_weak_typing)?

> However, there is no precise technical definition of what the terms mean

> Generally, a strongly typed language has stricter typing rules at compile time.

I don't think that a few implicit conversions amount to that big of a difference.

**Comment by Josh-Cena** [comment]
_Sigh_ Python doesn't have "compile time". Everything is runtime.

> I don't think that a few implicit conversions amount to that big of a difference.

Sure, you can argue that drawing the distinction between "strongly typed" and "weakly typed" is pretty much useless in actual production, but that's the entire point that's relevant to this conversation, about whether implicit coercions should be allowed in JS/TS.

**Comment by Josh-Cena** [comment]
Also, for folks in this thread interested about a lint rule: https://typescript-eslint.io/rules/restrict-plus-operands

**Comment by KotlinIsland** [comment]
> Python is designed to forbid implicit coercions

Python allows writing custom types that perform implicit conversions:
```py
class ImplicitStr(Str):
    def __add__(self, other):
        return super().__add__(self, str(other))

a = ImplicitStr("hi")
b = 10
print(a + b)  # hi10
```

It's not really a language level thing, only some small behavior choices of the builtin types.

> _Sigh_ Python doesn't have "compile time". Everything is runtime.

Python is compiled at load time:
```py
def foo():
    return 2
import dis
print(dis.dis(foo))
#  1           0 LOAD_CONST               1 (2)
#              2 RETURN_VALUE
```

**Comment by Josh-Cena** [comment]
Yeah, every language has to be compiled to _some_ native code at runtime, unless you run on some specially designed VM for your script language.

About operator overloads, as soon as you opt into them, you are basically writing custom functions. Due to the dynamic-ness of Python functions, it's impossible to constrain the types of what's passed in unless you do that yourself.

I don't even know what we are arguing here. It's a language feature that Python disallows mingling variables in operations while JS allows them—that's literally the definition of "strongly typed" you'll find in every textbook. Full stop.

**Comment by Josh-Cena** [comment]
TBH, I always _believed_ that `1 + "2"` is equal to `3` in JS, no idea where that illusion came from🤦‍♂️

**Comment by Josh-Cena** [comment]
> there is some fundamental difference between JS and Python's type system that necessitates allowing misleading, confusing and error prone operations.

Except... there is? At least prior to ES6, all JS operations are permissive. It's never ingrained in the language that some operation, albeit magical and error-prone, should throw an error.

> So why allow `number` + `string`?

That's, in my understanding, due to the balance between TypeScript's pursuit of balance between soundness and ergonomics.

`[1, 2] + 3` is not a well-defined operation. Both sides being coerced to strings is confusing to anyone coming from another language. e.g. In C, it would be shifting a pointer.

`number + string`, OTOH, is not only well-defined, it's also idiomatically used across all languages. Your `foo` is a contrived example, but in actual code, it's fairly common to write `"The count is" + count` (prior to template literals being a thing). In actual context where the semantics of `a` and `b` are clear, making mistakes should be rare.

If `1 + "2"` results in `3`, I will agree it's a huge footgun; but if it is a simple string concatenation, it will fit most people's intuition. Take, for example, the fact that TS also allows `const a = "1" + [2, 3];` This is also the reason given in https://github.com/microsoft/TypeScript/issues/16895

**Comment by KotlinIsland** [comment]
> Except... there is [a fundamental difference]?

These differences are not 'fundamental' in any way. They amount to specific method implementation details at best.

> It's never ingrained in the language that some operation, albeit magical and error-prone, should throw an error.

This is __identical__ across Python and JS. There is no part of Pythons type system or syntax that enforces coercion/non-coercion. (depending on what you actually mean by 'operation' here at least)

Yeah, I can understand that TypeScript is held back by JavaScript in the sense that `1 + "1"` can appeal to being 'more familiar' to the JS landscape and that there isn't the luxury of being a new language without any tech debt or backward compatibility obligations. But I feel like it's clear that the better design direction is explicit and intentional, rather than implicit and coercive.
